### ISSUE TITLE
Activation-time error handling

### ISSUE THREAD

**Issue by TheRealPiotrP** [dotnet/vscode-csharp#1840]
At a high level, this extension seems to be a wrapper around two components: O# and coreclrdbg. Its current implementation allows for the extension to load despite either O# or coreclrdbg not being available. This results in an ambiguous definition of activation success both for end-users and for contributors to this repo.

I would like to simplify the activation path by:
 - making activation boolean. Either everything is installed or activation fails.
 - Throw an error on failed activation. notify VS Code to consider the extension to be broken. report activation failures via VS Code's activation handling mechanism.

Do folks have concern about this change? @gregg-miskelly and @DustinCampbell likely have significant insights regarding the current design and may prefer to leave it unchanged.

**Comment by DustinCampbell** [comment]
> Either everything is installed or activation fails.

I'm sort of OK with this for OmniSharp but I'm not sure it would work for the debugger, which is not supported on all of the platforms that OmniSharp can run on.

**Comment by gregg-miskelly** [comment]
Two notes:
1. It is definitely useful to have the extension function in cases where OmniSharp can successfully start, but the debugger cannot. This is mostly for mono/unity scenarios where the user isn't going to be .NET Core debugging.
2. I am NOT aware of cases where we will silently fail to start from a user perspective. Failures are indicated with error messages either in the channel and/or `vscode.window.showErrorMessage`. This is better than throwing exceptions back to VS Code as we have more control over the experience.

**Comment by TheRealPiotrP** [comment]
Thanks, I hadn't understood that debugging does not work on all platforms. This makes more sense to me now.

I guess the thing I don't understand are the conditions under which `ensureRuntimeDependencies` will return `false` while we are still comfortable moving forward with execution? I had made the naiive assumption that this method will determine what downloads are applicable and will ensure that they are all downloaded. When this method returns `false`, it sounds like you are saying that O# is downloaded successfully while the debugger is not applicable or not downloaded. Is that an accurate interpretation? If not, I'm afraid I'm still confused about the state when `ensureRuntimeDependencies` is `false`...

@gregg-miskelly:

> 
It is definitely useful to have the extension function in cases where OmniSharp can successfully start, but the debugger cannot. This is mostly for mono/unity scenarios where the user isn't going to be .NET Core debugging.
> 

Agreed, since there are known scenarios where this is the expected behavior. However, I cannot see from the current code how we determine the expected behavior of the debugger. I'd be less confused if we could determine:
- Should the debugger be enabled?
- Are the debugger downloads available?
- Are the O# downloads available?

Given the current state in the activation method, it seems like we're using `ensureRuntimeDependencies` as a proxy for all of these decision points...

>
I am NOT aware of cases where we will silently fail to start from a user perspective. Failures are indicated with error messages either in the channel and/or vscode.window.showErrorMessage. This is better than throwing exceptions back to VS Code as we have more control over the experience.
>

This one is straightforward to repro. Open a `larger` project in vscode and, before the extension starts, issue the `.NET: Generate Assets for Build and Debug` command from the command palette. If O# has not finished spinning up yet then the command will fail with an error message, despite our extension claiming to be `activated`. It is not a silent failure, but it seems to be a misuse of the VSCode extensibility model. By registering this, and other, commands and completing activation we seem to be telling Code it's ok for users to be issuing our commands when in fact there is an initialization state during which we're not ready to honor that contract.

While writing this I realized I'm making some assumptions about VS Code which are driving my thinking and I need to validate these. Specifically, I'm assuming that if we surface errors and state changes to VS Code then we will provide our users with a more deterministic experience. Does VS Code make available commands from extensions that have not yet [or have failed to] activated? If not, our current strategy makes much more sense to me. If so, I think we'd be better citizens by failing to activate and removing those commands from the user's command palette. In this scenario I think that reporting errors with vscode.window.showErrorMessage is beneficial but insufficient for providing the expected end-to-end UX.

I'll do some experiments to disprove my assumptions.
